File: desktop/src/features/evidence-cards/lib/triageReviewCards.ts
import { sha256 } from "@noble/hashes/sha2.js";
import { bytesToHex } from "@noble/hashes/utils.js";
import { z } from "zod";

const httpUrlSchema = z
  .string()
  .max(2_048)
  .refine((value) => {
    try {
      return ["http:", "https:"].includes(new URL(value).protocol);
    } catch {
      return false;
    }
  });

const boundedText = (max: number) => z.string().trim().min(1).max(max);

export const TRIAGE_RISKS = ["low", "medium", "high"] as const;
export type TriageRisk = (typeof TRIAGE_RISKS)[number];

export const REVIEW_DECISIONS = [
  "pending",
  "approved",
  "changes_requested",
  "rejected",
] as const;
export type ReviewDecision = (typeof REVIEW_DECISIONS)[number];

const triageGroupSchema = z.object({
  label: boundedText(160),
  count: z.number().int().min(1),
  oldest_age: boundedText(64).optional(),
  risk: z.enum(TRIAGE_RISKS),
});

const triageCardPayloadSchema = z.object({
  schema_version: z.literal(1),
  triage_id: z.string().uuid(),
  title: boundedText(160),
  groups: z.array(triageGroupSchema).min(1).max(32),
  proposed_action: boundedText(2_000),
  record_url: httpUrlSchema.optional(),
  shadow: z.boolean(),
});

const reviewCardPayloadSchema = z.object({
  schema_version: z.literal(1),
  review_id: z.string().uuid(),
  title: boundedText(160),
  patch_ref: boundedText(2_048),
  evidence: z.array(boundedText(500)).max(16),
  test_summary: boundedText(2_000).optional(),
  decision: z.enum(REVIEW_DECISIONS),
  reviewer: boundedText(160).optional(),
  record_url: httpUrlSchema.optional(),
  shadow: z.boolean(),
});

export type TriageCardPayload = z.infer<typeof triageCardPayloadSchema>;
export type ReviewCardPayload = z.infer<typeof reviewCardPayloadSchema>;

export type ParsedTriageCard = {
  payload: TriageCardPayload;
  payloadHash: string;
};

export type ParsedReviewCard = {
  payload: ReviewCardPayload;
  payloadHash: string;
};

function findTag(tags: string[][], name: string): string | undefined {
  return tags.find((tag) => tag[0] === name)?.[1];
}

function verifyPayloadHash(
  encoded: string,
  payloadHash: string | undefined,
): string | null {
  if (!payloadHash || !/^[0-9a-f]{64}$/i.test(payloadHash)) {
    return null;
  }
  const encodedHash = bytesToHex(sha256(new TextEncoder().encode(encoded)));
  if (encodedHash !== payloadHash.toLowerCase()) {
    return null;
  }
  return encodedHash;
}

export function parseTriageCard(tags: string[][]): ParsedTriageCard | null {
  const encoded = findTag(tags, "triage_card");
  const payloadHash = findTag(tags, "payload_hash");
  if (!encoded) return null;

  const verifiedHash = verifyPayloadHash(encoded, payloadHash);
  if (!verifiedHash) return null;

  try {
    const payload = triageCardPayloadSchema.parse(JSON.parse(encoded));
    return { payload, payloadHash: verifiedHash };
  } catch {
    return null;
  }
}

export function parseReviewCard(tags: string[][]): ParsedReviewCard | null {
  const encoded = findTag(tags, "review_card");
  const payloadHash = findTag(tags, "payload_hash");
  if (!encoded) return null;

  const verifiedHash = verifyPayloadHash(encoded, payloadHash);
  if (!verifiedHash) return null;

  try {
    const payload = reviewCardPayloadSchema.parse(JSON.parse(encoded));
    return { payload, payloadHash: verifiedHash };
  } catch {
    return null;
  }
}

export function buildTriageCardContent(payload: TriageCardPayload): string {
  triageCardPayloadSchema.parse(payload);
  const groupLines = payload.groups.map((group) => {
    const age = group.oldest_age ? ` · oldest ${group.oldest_age}` : "";
    return `- ${group.label}: ${group.count}${age} (${group.risk})`;
  });
  const lines = [
    `## ${payload.title}`,
    groupLines.join("\n"),
    `**Proposed action:** ${payload.proposed_action}`,
  ];
  if (payload.shadow) {
    lines.push("**SHADOW / NOT DELIVERED**");
  }
  return lines.join("\n\n");
}

export function buildReviewCardContent(payload: ReviewCardPayload): string {
  reviewCardPayloadSchema.parse(payload);
  const labels: Record<ReviewDecision, string> = {
    pending: "⏳ Pending",
    approved: "✅ Approved",
    changes_requested: "✏️ Changes requested",
    rejected: "⛔ Rejected",
  };
  const lines = [
    `## ${payload.title}`,
    `${labels[payload.decision]} — ${payload.patch_ref}`,
  ];
  if (payload.evidence.length > 0) {
    lines.push(`**Evidence:** ${payload.evidence.join("; ")}`);
  }
  if (payload.test_summary) {
    lines.push(`**Tests:** ${payload.test_summary}`);
  }
  if (payload.reviewer) {
    lines.push(`**Reviewer:** ${payload.reviewer}`);
  }
  if (payload.shadow) {
    lines.push("**SHADOW / NOT DELIVERED**");
  }
  return lines.join("\n\n");
}

export function buildTriageCardTags(input: {
  channelId: string;
  payload: TriageCardPayload;
  recipientPubkeys?: string[];
  rootEventId?: string | null;
  parentEventId?: string | null;
}): string[][] {
  const parsed = triageCardPayloadSchema.parse(input.payload);
  const encoded = JSON.stringify(parsed);
  const payloadHash = bytesToHex(sha256(new TextEncoder().encode(encoded)));
  const tags: string[][] = [["h", input.channelId]];
  for (const pubkey of new Set(input.recipientPubkeys ?? [])) {
    if (pubkey) tags.push(["p", pubkey]);
  }
  if (input.rootEventId && input.parentEventId) {
    if (input.rootEventId !== input.parentEventId) {
      tags.push(["e", input.rootEventId, "", "root"]);
    }
    tags.push(["e", input.parentEventId, "", "reply"]);
  }
  tags.push(
    ["triage_card", encoded],
    ["payload_hash", payloadHash],
    ["shadow", parsed.shadow ? "1" : "0"],
  );
  return tags;
}

export function buildReviewCardTags(input: {
  channelId: string;
  payload: ReviewCardPayload;
  recipientPubkeys?: string[];
  rootEventId?: string | null;
  parentEventId?: string | null;
}): string[][] {
  const parsed = reviewCardPayloadSchema.parse(input.payload);
  const encoded = JSON.stringify(parsed);
  const payloadHash = bytesToHex(sha256(new TextEncoder().encode(encoded)));
  const tags: string[][] = [["h", input.channelId]];
  for (const pubkey of new Set(input.recipientPubkeys ?? [])) {
    if (pubkey) tags.push(["p", pubkey]);
  }
  if (input.rootEventId && input.parentEventId) {
    if (input.rootEventId !== input.parentEventId) {
      tags.push(["e", input.rootEventId, "", "root"]);
    }
    tags.push(["e", input.parentEventId, "", "reply"]);
  }
  tags.push(
    ["review_card", encoded],
    ["payload_hash", payloadHash],
    ["shadow", parsed.shadow ? "1" : "0"],
  );
  return tags;
}

File: crates/buzz-sdk/tests/triage_review_cards.rs
use buzz_core::kind::{KIND_STREAM_REVIEW_CARD, KIND_STREAM_TRIAGE_CARD};
use buzz_core::review_card::{ReviewCardPayload, ReviewDecision};
use buzz_core::triage_card::{TriageCardPayload, TriageGroup, TriageRisk};
use buzz_sdk::{build_review_card, build_triage_card, ThreadRef};
use nostr::{EventBuilder, EventId, Keys};
use uuid::Uuid;

fn sign(builder: EventBuilder) -> nostr::Event {
    builder
        .sign_with_keys(&Keys::generate())
        .expect("test event should sign")
}

fn tag_value<'a>(event: &'a nostr::Event, name: &str) -> Option<&'a str> {
    event.tags.iter().find_map(|tag| {
        let parts = tag.as_slice();
        (parts.first().map(String::as_str) == Some(name))
            .then(|| parts.get(1).map(String::as_str))
            .flatten()
    })
}

#[test]
fn triage_card_keeps_markdown_fallback_and_signed_structured_payload() {
    let channel_id = Uuid::new_v4();
    let payload = TriageCardPayload {
        schema_version: 1,
        triage_id: Uuid::new_v4(),
        title: "68 shadow drafts need triage".into(),
        groups: vec![
            TriageGroup {
                label: "High risk · >7d".into(),
                count: 12,
                oldest_age: Some("14d".into()),
                risk: TriageRisk::High,
            },
            TriageGroup {
                label: "Medium risk · sales".into(),
                count: 41,
                oldest_age: Some("3d".into()),
                risk: TriageRisk::Medium,
            },
        ],
        proposed_action: "Open the oldest high-risk group first.".into(),
        record_url: Some("https://stomaton.example/triage/68".into()),
        shadow: true,
    };
    let fallback = "## Triage\n68 shadow drafts need triage\n\n**SHADOW — NOT DELIVERED**";

    let event = sign(
        build_triage_card(channel_id, &payload, fallback, None)
            .expect("valid triage card should build"),
    );

    assert_eq!(event.kind.as_u16(), KIND_STREAM_TRIAGE_CARD as u16);
    assert_eq!(event.content, fallback);
    assert_eq!(
        tag_value(&event, "h"),
        Some(channel_id.to_string().as_str())
    );
    assert_eq!(tag_value(&event, "shadow"), Some("1"));

    let encoded = tag_value(&event, "triage_card").expect("triage_card tag");
    let decoded: TriageCardPayload = serde_json::from_str(encoded).expect("valid payload JSON");
    assert_eq!(decoded, payload);

    let payload_hash = tag_value(&event, "payload_hash").expect("payload_hash tag");
    assert_eq!(payload_hash, payload.payload_hash().expect("payload hash"));
    assert_eq!(payload_hash.len(), 64);
}

#[test]
fn review_card_keeps_markdown_fallback_and_thread_refs() {
    let channel_id = Uuid::new_v4();
    let root_id = EventId::from_hex(&"11".repeat(32)).expect("root event id");
    let parent_id = EventId::from_hex(&"22".repeat(32)).expect("parent event id");
    let payload = ReviewCardPayload {
        schema_version: 1,
        review_id: Uuid::new_v4(),
        title: "Evidence card suite slice 3".into(),
        patch_ref: "https://github.com/block/buzz/pull/999".into(),
        evidence: vec!["Unit tests cover hash stability.".into()],
        test_summary: Some("cargo test -p buzz-core: ok".into()),
        decision: ReviewDecision::ChangesRequested,
        reviewer: Some("tyler".into()),
        record_url: None,
        shadow: true,
    };
    let fallback = "## Review\nChanges requested — SHADOW / NOT DELIVERED";
    let thread_ref = ThreadRef {
        root_event_id: root_id,
        parent_event_id: parent_id,
    };

    let event = sign(
        build_review_card(channel_id, &payload, fallback, Some(&thread_ref))
            .expect("valid review card should build"),
    );

    assert_eq!(event.kind.as_u16(), KIND_STREAM_REVIEW_CARD as u16);
    assert_eq!(event.content, fallback);
    assert_eq!(
        tag_value(&event, "h"),
        Some(channel_id.to_string().as_str())
    );
    assert_eq!(tag_value(&event, "shadow"), Some("1"));

    let encoded = tag_value(&event, "review_card").expect("review_card tag");
    let decoded: ReviewCardPayload = serde_json::from_str(encoded).expect("valid review JSON");
    assert_eq!(decoded, payload);

    let payload_hash = tag_value(&event, "payload_hash").expect("payload_hash tag");
    assert_eq!(payload_hash, payload.payload_hash().expect("payload hash"));

    let e_tags: Vec<Vec<String>> = event
        .tags
        .iter()
        .filter(|tag| tag.as_slice().first().map(String::as_str) == Some("e"))
        .map(|tag| tag.as_slice().to_vec())
        .collect();
    assert_eq!(
        e_tags[0],
        vec!["e".into(), root_id.to_hex(), "".into(), "root".into()]
    );
    assert_eq!(
        e_tags[1],
        vec!["e".into(), parent_id.to_hex(), "".into(), "reply".into()]
    );
}

#[test]
fn triage_card_rejects_empty_groups() {
    let payload = TriageCardPayload {
        schema_version: 1,
        triage_id: Uuid::new_v4(),
        title: "Triage".into(),
        groups: vec![],
        proposed_action: "Hold.".into(),
        record_url: None,
        shadow: true,
    };

    assert!(build_triage_card(Uuid::new_v4(), &payload, "fallback", None).is_err());
}

#[test]
fn review_card_rejects_non_http_record_url() {
    let payload = ReviewCardPayload {
        schema_version: 1,
        review_id: Uuid::new_v4(),
        title: "Review".into(),
        patch_ref: "abc123".into(),
        evidence: vec![],
        test_summary: None,
        decision: ReviewDecision::Pending,
        reviewer: None,
        record_url: Some("javascript:alert(1)".into()),
        shadow: true,
    };

    assert!(build_review_card(Uuid::new_v4(), &payload, "fallback", None).is_err());
}
